In [1]:
import pandas as pd
import json

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

The files we are using are from the project MTGJSON (https://mtgjson.com/), and are actively maintained, so we won't expect much in the way of missing or incorrect information.  However, the original format includes a lot of extranneous information, and we'll need to do several different transformations of datatypes and similar small formatting in order to get each dataframe into the shape that we want for Exploratory Data Analysis and, eventually, model building.  I'll be downloading individual `.json` files for each set that will be included in our analysis, so we will start by looking at what is necessary to clean one file, then create a function that will allow us to quickly and easily clean up any additional sets.


Each `.json` file is named via a three-character set abbreviation.  We'll use the Adventures in the Forgotten Realms (AFR) set to do our initial step-by-step cleaning.

In [2]:
df = pd.read_json('datasets/AFR.json', orient='index')
df

date         version  baseSetSize  \
meta 2022-06-20  5.2.0+20220620          NaN   
data        NaT             NaN        281.0   

                                                booster  \
meta                                                NaN   
data  {'arena': {'boosters': [{'contents': {'common'...   

                                                  cards  cardsphereSetId code  \
meta                                                NaN              NaN  NaN   
data  [{'artist': 'Jarel Threat', 'availability': ['...           1337.0  AFR   

      isFoilOnly  isOnlineOnly keyruneCode   mcmId  mcmIdExtras  \
meta         NaN           NaN         NaN     NaN          NaN   
data         0.0           0.0         AFR  4259.0       4260.0   

                                 mcmName mtgoCode  \
meta                                 NaN      NaN   
data  Adventures in the Forgotten Realms      AFR   

                                    name releaseDate  \
meta                                 NaN         NaN   
data  Adventures in the Forgotten Realms  2021-07-23   

                                          sealedProduct  tcgplayerGroupId  \
meta                                                NaN               NaN   
data  [{'identifiers': {'tcgplayerProductId': '23873...            2823.0   

                                                 tokens  totalSetSize  \
meta                                                NaN           NaN   
data  [{'artist': 'Irina Nordsol', 'availability': [...         402.0   

     translations       type  
meta          NaN        NaN  
data           {}  expansion

The initial dataframe consists of only two rows, "meta" and "data."  All of the data that we're currently interested in can be found in the cards column of the data row, so we will extract that as its own dataframe.

In [3]:
df = pd.DataFrame(df.cards[1])

In [4]:
df.head()

artist          availability boosterTypes borderColor  \
0      Jarel Threat  [arena, mtgo, paper]      [draft]       black   
1       Justyna Gil  [arena, mtgo, paper]      [draft]       black   
2   Oriana Menendez  [arena, mtgo, paper]      [draft]       black   
3  Daniel Ljunggren  [arena, mtgo, paper]      [draft]       black   
4     Johannes Voss  [arena, mtgo, paper]      [draft]       black   

  colorIdentity colors  convertedManaCost  edhrecRank         finishes  \
0           [W]    [W]                2.0     11959.0  [nonfoil, foil]   
1           [W]    [W]                4.0     16970.0  [nonfoil, foil]   
2           [W]    [W]                3.0     12876.0  [nonfoil, foil]   
3           [W]    [W]                3.0      3378.0  [nonfoil, foil]   
4           [W]    [W]                3.0      7131.0  [nonfoil, foil]   

                                          flavorText  \
0  The weight of this magic weapon falls heavy on...   
1  The noble steeds of the elven gods can sometim...   
2  Blink dogs protect the wild places of the worl...   
3                                                NaN   
4            When all hope fails, its light remains.   

                                         foreignData frameVersion  hasFoil  \
0                                                 []         2015     True   
1  [{'flavorText': 'Die edlen Reittiere der elfis...         2015     True   
2  [{'flavorText': 'Flimmerhunde beschützen die w...         2015     True   
3  [{'language': 'German', 'multiverseId': 527551...         2015     True   
4  [{'flavorText': 'Wenn alle Hoffnung schwindet,...         2015     True   

   hasNonFoil                                        identifiers  \
0        True  {'cardKingdomFoilId': '248423', 'cardKingdomId...   
1        True  {'cardKingdomFoilId': '248542', 'cardKingdomId...   
2        True  {'cardKingdomFoilId': '248564', 'cardKingdomId...   
3        True  {'cardKingdomFoilId': '248511', 'cardKingdomId...   
4        True  {'cardKingdomFoilId': '248369', 'cardKingdomId...   

          keywords language  layout  \
0          [Equip]  English  normal   
1         [Flying]  English  normal   
2  [Double strike]  English  normal   
3              NaN  English  normal   
4              NaN  English  normal   

                                          legalities   manaCost  manaValue  \
0  {'alchemy': 'Legal', 'brawl': 'Legal', 'comman...     {1}{W}        2.0   
1  {'alchemy': 'Legal', 'brawl': 'Legal', 'comman...     {3}{W}        4.0   
2  {'alchemy': 'Legal', 'brawl': 'Legal', 'comman...     {2}{W}        3.0   
3  {'alchemy': 'Legal', 'brawl': 'Legal', 'comman...  {W}{W}{W}        3.0   
4  {'alchemy': 'Legal', 'brawl': 'Legal', 'comman...     {2}{W}        3.0   

                        name number    printings  \
0                    +2 Mace      1        [AFR]   
1            Arborea Pegasus      2        [AFR]   
2                  Blink Dog      3        [AFR]   
3  The Book of Exalted Deeds      4  [AFR, PAFR]   
4          Celestial Unicorn      5        [AFR]   

                                        purchaseUrls    rarity  \
0  {'cardKingdom': 'https://mtgjson.com/links/246...    common   
1  {'cardKingdom': 'https://mtgjson.com/links/d46...    common   
2  {'cardKingdom': 'https://mtgjson.com/links/d18...  uncommon   
3  {'cardKingdom': 'https://mtgjson.com/links/eba...    mythic   
4  {'cardKingdom': 'https://mtgjson.com/links/7c6...    common   

                                             rulings setCode     subtypes  \
0                                                 []     AFR  [Equipment]   
1                                                 []     AFR    [Pegasus]   
2  [{'date': '2021-07-23', 'text': 'Phased out pe...     AFR        [Dog]   
3  [{'date': '2021-07-23', 'text': 'The enlighten...     AFR           []   
4  [{'date': '2021-07-23', 'text': 'If you gain l...     AFR    [Unicorn]   

    supertypes                                         

In [5]:
df.shape

(420, 51)

We can see from our row count that we have 420 different cards represented, however a quick reference of the List of MTG Sets article on Wikipedia (https://en.wikipedia.org/wiki/List_of_Magic:_The_Gathering_sets#Expansion_sets) tells us we should expect to see only 281 cards.  The MTGJSON files include separate entries for promotional versions of cards with alternate art as well as digital-only versions of cards that won't be important to our analysis, so we'll drop these from the list.

In [6]:
df = df[df['promoTypes'].isnull()]
df.shape

(299, 51)

In [7]:
df = df.loc[df['securityStamp'] != 'arena']
df.shape

(281, 51)

281 cards is exactly what we expect to see, so now we will drop the Basic Lands from the dataset.  Every expansion includes the five basic land cards that are necessary to play the game, but they are in and of themselves irrelevant to any analysis that we plan to do, so we can drop them from each of the sets that we include in our analysis.

In [8]:
df = df.loc[~df['supertypes'].map(lambda x: bool('Basic' in x))]
df.shape

(261, 51)

The MTGJSON files include a wealth of information about each card, most of which will not be germaine to this specific analysis, so we choose a number that should be relevant, and drop the rest.

In [9]:
df = df.loc[:, ['name', 'colors', 'manaValue', 'manaCost', 'power', 'toughness', 
                    'loyalty', 'text', 'rarity', 'keywords', 'supertypes', 'types']]

In [10]:
df.head(10)

name colors  manaValue   manaCost power toughness  \
0                     +2 Mace    [W]        2.0     {1}{W}   NaN       NaN   
1             Arborea Pegasus    [W]        4.0     {3}{W}     2         3   
2                   Blink Dog    [W]        3.0     {2}{W}     1         1   
3   The Book of Exalted Deeds    [W]        3.0  {W}{W}{W}   NaN       NaN   
4           Celestial Unicorn    [W]        3.0     {2}{W}     3         2   
5                Cleric Class    [W]        1.0        {W}   NaN       NaN   
6           Cloister Gargoyle    [W]        3.0     {2}{W}     0         4   
8               Dancing Sword    [W]        2.0     {1}{W}   NaN       NaN   
9          Dawnbringer Cleric    [W]        2.0     {1}{W}     1         3   
10             Delver's Torch    [W]        2.0     {1}{W}   NaN       NaN   

   loyalty                                               text    rarity  \
0      NaN  Equipped creature gets +2/+2.\nEquip {3} ({3}:...    common   
1      NaN  Flying\nWhen Arborea Pegasus enters the battle...    common   
2      NaN  Double strike\nTeleport — {3}{W}: Blink Dog ph...  uncommon   
3      NaN  At the beginning of your end step, if you gain...    mythic   
4      NaN  Whenever you gain life, put a +1/+1 counter on...    common   
5      NaN  (Gain the next level as a sorcery to add its a...  uncommon   
6      NaN  When Cloister Gargoyle enters the battlefield,...  uncommon   
8      NaN  Equipped creature gets +2/+1.\nWhen equipped c...      rare   
9      NaN  When Dawnbringer Cleric enters the battlefield...    common   
10     NaN  Equipped creature gets +1/+1.\nWhenever equipp...    common   

                             keywords   supertypes                 types  
0                             [Equip]           []            [Artifact]  
1                            [Flying]           []            [Creature]  
2                     [Double strike]           []            [Creature]  
3                                 NaN  [Legendary]            [Artifact]  
4                                 NaN           []            [Creature]  
5                                 NaN           []         [Enchantment]  
6          [Venture into the dungeon]           []  [Artifact, Creature]  
8                             [Equip]           []            [Artifact]  
9                                 NaN           []            [Creature]  
10  [Equip, Venture into the dungeon]           []            [Artifact]

Several columns of information that will be important to us are formatted as lists, which makes getting to the information with Pandas a bit tricky, so we're going to convert them from lists to strings.  We'll do this for the columns "colors", "keywords", "supertypes", and "types".  Unfortunately, the keywords column has NaN values instead of empty lists if the card has no keywords, so we'll first replace those NaNs with empty strings.

In [11]:
df['keywords'].fillna('', inplace=True)
to_unlist = ['colors', 'types', 'supertypes', 'keywords']
for column in to_unlist:
    df[column] = [', '.join(map(str, l)) for l in df[column]]

In [12]:
df.head(10)

name colors  manaValue   manaCost power toughness  \
0                     +2 Mace      W        2.0     {1}{W}   NaN       NaN   
1             Arborea Pegasus      W        4.0     {3}{W}     2         3   
2                   Blink Dog      W        3.0     {2}{W}     1         1   
3   The Book of Exalted Deeds      W        3.0  {W}{W}{W}   NaN       NaN   
4           Celestial Unicorn      W        3.0     {2}{W}     3         2   
5                Cleric Class      W        1.0        {W}   NaN       NaN   
6           Cloister Gargoyle      W        3.0     {2}{W}     0         4   
8               Dancing Sword      W        2.0     {1}{W}   NaN       NaN   
9          Dawnbringer Cleric      W        2.0     {1}{W}     1         3   
10             Delver's Torch      W        2.0     {1}{W}   NaN       NaN   

   loyalty                                               text    rarity  \
0      NaN  Equipped creature gets +2/+2.\nEquip {3} ({3}:...    common   
1      NaN  Flying\nWhen Arborea Pegasus enters the battle...    common   
2      NaN  Double strike\nTeleport — {3}{W}: Blink Dog ph...  uncommon   
3      NaN  At the beginning of your end step, if you gain...    mythic   
4      NaN  Whenever you gain life, put a +1/+1 counter on...    common   
5      NaN  (Gain the next level as a sorcery to add its a...  uncommon   
6      NaN  When Cloister Gargoyle enters the battlefield,...  uncommon   
8      NaN  Equipped creature gets +2/+1.\nWhen equipped c...      rare   
9      NaN  When Dawnbringer Cleric enters the battlefield...    common   
10     NaN  Equipped creature gets +1/+1.\nWhenever equipp...    common   

                           keywords supertypes               types  
0                             Equip                       Artifact  
1                            Flying                       Creature  
2                     Double strike                       Creature  
3                                    Legendary            Artifact  
4                                                         Creature  
5                                                      Enchantment  
6          Venture into the dungeon             Artifact, Creature  
8                             Equip                       Artifact  
9                                                         Creature  
10  Equip, Venture into the dungeon                       Artifact

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 261 entries, 0 to 278
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        261 non-null    object 
 1   colors      261 non-null    object 
 2   manaValue   261 non-null    float64
 3   manaCost    252 non-null    object 
 4   power       142 non-null    object 
 5   toughness   142 non-null    object 
 6   loyalty     5 non-null      object 
 7   text        261 non-null    object 
 8   rarity      261 non-null    object 
 9   keywords    261 non-null    object 
 10  supertypes  261 non-null    object 
 11  types       261 non-null    object 
dtypes: float64(1), object(11)
memory usage: 26.5+ KB


We can see that there are four numeric columns that will likely interest us:  manaValue, power, toughness & loyalty.  However, the latter three are currently formatted as strings, so we will fix that. There is the possibility that values in these lists may include the value '*', indicating that they value is variable.  For simplicity's sake, we will use `errors='coerce'` to replace the variable power indicator with NaN.

In [14]:
to_convert = ['power', 'toughness', 'loyalty']
for column in to_convert:
    df[column] = pd.to_numeric(df[column], errors='coerce')

We've removed some items from the set, so we'll go ahead and reset the index numbers.

In [15]:
df.reset_index(drop=True, inplace=True)

In [16]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 261 entries, 0 to 260
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        261 non-null    object 
 1   colors      261 non-null    object 
 2   manaValue   261 non-null    float64
 3   manaCost    252 non-null    object 
 4   power       142 non-null    float64
 5   toughness   142 non-null    float64
 6   loyalty     5 non-null      float64
 7   text        261 non-null    object 
 8   rarity      261 non-null    object 
 9   keywords    261 non-null    object 
 10  supertypes  261 non-null    object 
 11  types       261 non-null    object 
dtypes: float64(4), object(8)
memory usage: 24.6+ KB


This looks like a good place to start from, so we'll create a function that includes all of these steps.  We can then put each of the individual sets that we want to clean in a list and iterate over it with the function to give us a list of dataframes, which we can then write out to `.csv` format to give us a number of cleaned individual datasets that are identically formatted, which will allow us to easily combine them.

#### Handling Modal Cards

Many cards in the sets that we are considering are modal in nature, meaning that a single card can be played in one of two distinct ways.  In the MTGJSON datasets, each mode of a single card has its own row, but both rows have the same name in the format "MODE 1 // MODE 2". For the sake of this analysis, we will treat each mode of such a card as its own card. We will add a function to be called within our cleaning function that will rename each of these entries to reflect the name of casting that specific mode.

In [17]:
def drop_modal_names (df):
    indexes = df.loc[df.name.str.contains('//'), ['name']].index
    for count, value in enumerate(df.loc[df.name.str.contains('//'), ['name']].name):
        if count % 2 == 0:
            c = value.find('/') - 1
            df.iat[indexes[count], 0] = value[:c]
        else:
            c = value.find('/') + 3
            df.iat[indexes[count], 0] = value[c:]

In [18]:
def initial_cleaning(json_file):
    df = pd.read_json(json_file, orient='index')
    df = pd.DataFrame(df.cards[1])
    # Removing promos, online only cards, basic lands.    
    df = df[df['promoTypes'].isnull()]
    df = df.loc[df['securityStamp'] != 'arena']
    df = df.loc[~df['supertypes'].map(lambda x: bool('Basic' in x))]
    # Choosing the columns we want to keep.
    df = df.loc[:, ['name', 'colors', 'manaValue', 'manaCost', 'power', 'toughness', 
                    'loyalty', 'text', 'rarity', 'keywords', 'supertypes', 'types']]
    # Resetting the index before we do anything else.
    df.reset_index(drop=True, inplace=True)
    # Changing list values to string values
    to_unlist = ['colors', 'types', 'supertypes', 'keywords']
    df['keywords'].fillna('', inplace=True)
    for column in to_unlist:
        df[column] = [', '.join(map(str, l)) for l in df[column]]
    # Converting numeric columns to numeric datatypes
    to_convert = ['power', 'toughness', 'loyalty']
    for column in to_convert:
        df[column] = pd.to_numeric(df[column], errors='coerce')
    # Renaming modal cards
    drop_modal_names(df)
    # Just in case we missed any duplicated entries with our initial cleaning steps, we drop them here and
    # reset the index one last time.
    df.drop_duplicates(inplace=True)
    df.reset_index(drop=True, inplace=True)
    return df

In [19]:
files = ['AFR', 'ELD', 'IKO', 'KHM', 'MID', 'NEO', 'SNC', 'STX', 'THB', 'VOW', 'ZNR']
dfs = []

In [20]:
for f in files:
    dfs.append(initial_cleaning('datasets/'+f+'.json'))

In [21]:
for df, f in zip(dfs, files):
    df.to_csv('cleaned_datasets/'+f+'_cleaned.csv')

With our individual sets cleaned and exported, we'll now merge them all in to a single dataframe, using the `reduce()` function.

In [22]:
from functools import reduce

In [23]:
df_merged = reduce(lambda left,right: pd.merge(left, right, how='outer'), dfs)

In [24]:
df_merged.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 3119 entries, 0 to 3118
Data columns (total 12 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   name        3119 non-null   object 
 1   colors      3119 non-null   object 
 2   manaValue   3119 non-null   float64
 3   manaCost    2826 non-null   object 
 4   power       1667 non-null   float64
 5   toughness   1672 non-null   float64
 6   loyalty     44 non-null     float64
 7   text        3095 non-null   object 
 8   rarity      3119 non-null   object 
 9   keywords    3119 non-null   object 
 10  supertypes  3119 non-null   object 
 11  types       3119 non-null   object 
dtypes: float64(4), object(8)
memory usage: 316.8+ KB


In [25]:
df_merged[df_merged.duplicated(subset= 'name', keep=False)]

Empty DataFrame
Columns: [name, colors, manaValue, manaCost, power, toughness, loyalty, text, rarity, keywords, supertypes, types]
Index: []

In [26]:
df_merged.describe()

manaValue        power    toughness    loyalty
count  3119.000000  1667.000000  1672.000000  44.000000
mean      2.943892     2.659868     2.898923   4.045455
std       1.620426     1.551854     1.570204   0.861436
min       0.000000     0.000000     0.000000   2.000000
25%       2.000000     2.000000     2.000000   3.750000
50%       3.000000     2.000000     3.000000   4.000000
75%       4.000000     3.000000     4.000000   5.000000
max      15.000000    11.000000    17.000000   5.000000

In [27]:
df_merged.shape

(3119, 12)

In [28]:
df_merged.to_csv('cleaned_datasets/allsets_cleaned.csv')